TombEditor/EditorActions.cs

```diff
@@ -465,6 +465,103 @@ public static void SmoothSector(Room room, int x, int z, SectorVerticalPart vert
             SmartBuildGeometry(room, new RectangleInt2(x, z, x, z));
         }
 
+        public static void SmoothTerrain(Room room, RectangleInt2 area, SectorVerticalPart vertical, bool disableUndo = false)
+        {
+            if (!disableUndo)
+            {
+                HashSet<Room> affectedRooms = room.GetAdjoiningRoomsFromArea(area);
+                affectedRooms.Add(room);
+                _editor.UndoManager.PushGeometryChanged(affectedRooms);
+            }
+
+            bool isBorder;
+
+            // Iterate through the selected region (excluding border sectors)
+            for (int z = area.Y0; z <= area.Y1; z++)
+            {
+                isBorder = z <= 1 || z >= room.NumZSectors - 2;
+
+                if (isBorder)
+                    continue;
+
+                for (int x = area.X0; x <= area.X1; x++)
+                {
+                    isBorder = x <= 1 || x >= room.NumXSectors - 2;
+
+                    if (isBorder)
+                        continue;
+
+                    // Smooth all four corners of current sector
+                    for (SectorEdge edge = 0; edge < SectorEdge.Count; edge++)
+                        SmoothSectorCorner(room, x, z, edge, vertical);
+                }
+            }
+
+            SmartBuildGeometry(room, area);
+        }
+
+        private static void SmoothSectorCorner(Room room, int x, int z, SectorEdge edge, SectorVerticalPart vertical)
+        {
+            // Define the neighboring sectors and their corresponding edges based on the corner we're processing
+            (int X, int Z, SectorEdge Edge)[] neighbors = edge switch
```

Comment:
Consider turning this into a record struct.

Have you considered using a stack allocated Span for this?

Reply:
It's not a very re-usable tuple, so creating a struct for it seems rather wasteful. Also not sure if a span would improve anything here, as each collection is only 3 items.